## Convolutional Neural Network (CNN)

- CNN is basically two dimensional configuration of neural networks

- The input of CNN are image (3 N by N if it color image and N by N if its black and white image)

- The weights are also two dimensional array

<img src="rgb_image.png" width="300" height="300">

### The weights in CNN

- The weights in CNN are called:

    - Kernel 
    
    or    
    
    - Filter matrix
    
    
<img src="kernel_image.png" width="400" height="400">

### Stride

- To define a CNN, we should specify the horizontal and vertical movement steps 

- what is the output size with stride = 1 and stride =2?

<img src="stride_1.png" width="400" height="400">

<img src="stride_2.png" width="400" height="400">

- output_size = (input_size - filter_size)/stride + 1

- Stride visualization: http://deeplearning.stanford.edu/wiki/index.php/Feature_extraction_using_convolution

In [ ]:
model.add(Conv2D(32, kernel_size=(3, 3), strides=(1, 1), activation='relu', input_shape=input_shape))

### Max Pooling vs Average Pooling

- Max pooling: take the maximum element from each window of a certain size

<img src="maxpooling.png" width="400" height="400">

### Faltten Layer 

- After feature extraction that is done by multiple Convolutional layers, we use flatten layer to add MLP after convolutional layers in order to do classification task

- This one is simple--it's just Keras's verison of numpy.reshape. This reshapes n-dimensional arrays to a vector. This is necessary when moving from Conv2D layers, which expect 2-dimensional arrays as inputs, to Dense layers, which expect 1-dimension vectors as inputs. As a concrete example, a Flatten layer given a 28 x 28 array as input would output a vector of the shape (784, 1)

### Visualize the whole NN: CNN + MLP

<img src="CNN.png" width="600" height="600">

## Activity: Obtain the number of parameters for the following CNN

- By default, the strides = (1, 1)

In [1]:
from __future__ import print_function
import keras

from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D


model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(28, 28, 1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
# model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
# model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

print(model.summary())

/Users/miladtoutounchian/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               1179776   
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 1,199,882
Trainable params: 1,199,882
Non-trainable params: 0
_________________________________________________________________


### Solution

- output_size = (28 - 3)/1 + 1 = 26

- output_size = (26 - 3)/1 + 1 = 24

- The parameters for the first Conv2D = 32 x 9 + 32 = 320

- The parameters for the second Conv2D = 64 x 32 x 9 + 64 = 18496

- The shape for flatten is: 12 x 12 x 64 = 9216

- The parameters for dense_1 = 9216 x 128 + 128 = 1179776

- The parameters for dense_2 = 128 x 10 + 10 = 1290

## Data Preparation for CNN

- Suppose we want to feed a 4 by 4 image to a CNN network, how we should reshape the data?

In [1]:
from keras.layers import Input, Dense, Conv2D, MaxPooling2D, UpSampling2D
from keras.models import Model
from keras import backend as K
import numpy as np

input_img = Input(shape=(4, 4, 1))  # adapt this if using `channels_first` image data format

x = Conv2D(2, (2, 2), activation='relu')(input_img)
y = Conv2D(3, (2, 2), activation='relu')(x)
model = Model(input_img, y)
# cnv_ml_1 = Model(input_img, x)

data = np.array([[5, 12, 1, 8], [2, 10, 3, 6], [4, 7, 9, 1], [5, 7, 5, 6]])
data = data.reshape(4, 4, 1)
print(data[0])
data = np.expand_dims(data, axis=0)
# data = data.reshape(4, 4, 1)
print(data[0])
print(model.predict(data))
print('M :')
print(model.predict(data).reshape(3, 2, 2))
print(model.summary())

/Users/miladtoutounchian/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


[[ 5]
 [12]
 [ 1]
 [ 8]]
[[[ 5]
  [12]
  [ 1]
  [ 8]]

 [[ 2]
  [10]
  [ 3]
  [ 6]]

 [[ 4]
  [ 7]
  [ 9]
  [ 1]]

 [[ 5]
  [ 7]
  [ 5]
  [ 6]]]
[[[[0.         0.30877003 0.        ]
   [0.         1.240459   0.        ]]

  [[0.         0.         0.        ]
   [0.83486694 1.7766566  0.        ]]]]
M :
[[[0.         0.30877003]
  [0.         0.        ]]

 [[1.240459   0.        ]
  [0.         0.        ]]

 [[0.         0.83486694]
  [1.7766566  0.        ]]]
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 4, 4, 1)           0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 3, 3, 2)           10        
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 2, 2, 3)           27        
Total params: 37
Trainable params: 37
Non-trainable params: 0
________

In [2]:
from keras.layers import Input, Dense, Conv2D, MaxPooling2D, UpSampling2D
from keras.models import Model
from keras import backend as K
import numpy as np

input_img = Input(shape=(4, 4, 1))  # adapt this if using `channels_first` image data format

x = Conv2D(2, (2, 2), activation='relu')(input_img)
y = Conv2D(3, (2, 2), activation='relu')(x)
model = Model(input_img, y)
# cnv_ml_1 = Model(input_img, x)

data = np.array([[5, 12, 1, 8], [2, 10, 3, 6], [4, 7, 9, 1], [5, 7, 5, 6]])
data = data.reshape(1, 4, 4, 1)
print(model.predict(data))
print('M :')
print(model.predict(data).reshape(3, 2, 2))
print(model.summary())

[[[[0.         3.6707969  0.        ]
   [0.         0.         3.1801198 ]]

  [[0.         1.1546224  0.6197593 ]
   [0.         0.13880396 1.9919777 ]]]]
M :
[[[0.         3.6707969 ]
  [0.         0.        ]]

 [[0.         3.1801198 ]
  [0.         1.1546224 ]]

 [[0.6197593  0.        ]
  [0.13880396 1.9919777 ]]]
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 4, 4, 1)           0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 2)           10        
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 2, 2, 3)           27        
Total params: 37
Trainable params: 37
Non-trainable params: 0
_________________________________________________________________
None


In [3]:
from keras.datasets import mnist

img_rows, img_cols = 28, 28 

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)
    
print(x_train[0].shape)

(28, 28, 1)


In [4]:
print(x_train[1].shape)

(28, 28, 1)
